tests/tests/test_compat.py

```diff
@@ -0,0 +1,295 @@
+        if len(dev_invs) < num_devices:
+            time.sleep(1)
+            continue
+        # Check if inventories has been updated since starting this loop
+        num_updated = 0
+        for device in dev_invs:
+            # datetime does not have an RFC3339 parser, but we can convert
+            # the timestamp to a compatible ISO format by removing
+            # fractions and replacing the Zulu timezone with GMT.
+            updated_ts = datetime.fromisoformat(
+                device["updated_ts"].split(".")[0] + "+00:00"
+            )
+            if updated_ts > update_after:
+                num_updated += 1
+            else:
+                time.sleep(1)
+                break
+
+
+def assert_successful_deployment(api_deployments, deployment_id, timeout=TIMEOUT):
+    """
+    Waits for the ongoing deployment (specified by deployment_id) to finish
+    and asserting all devices were successfully upgraded.
+    :param api_deployments: testutils.api.client.ApiClient setup and authorized
+                             to use the deployments management api,
+                             i.e. api_client.with_auth(api_token)
+    :param deployment_id: deployment id to watch
+    :param timeout: optional timeout value to wait for deployment (defaults to 5min)
+    """
+    deadline = datetime.now() + timeout
+    while True:
+        rsp = api_deployments.call(
+            "GET", deployments.URL_DEPLOYMENTS_ID.format(id=deployment_id)
+        )
+        assert rsp.status_code == 200
+
+        dpl = rsp.json()
+        if dpl["status"] == "finished":
+            rsp = api_deployments.call(
+                "GET", deployments.URL_DEPLOYMENTS_STATISTICS.format(id=deployment_id),
+            )
+            assert rsp.status_code == 200
+            assert rsp.json()["failure"] == 0
+            assert rsp.json()["success"] == dpl["device_count"]
+            break
+        elif datetime.now() > deadline:
+            pytest.fail("timeout: Waiting for devices to update")
+        else:
+            time.sleep(1)
+
+
+class TestClientCompatibilityBase:
+    """
+    This class contains compatibility tests implementation for assessing
+    server compatibility with older clients.
+    """
+
+    def compatibility_test_impl(self, env):
+        """
+        The actual test implementation:
+         - Accept devices
+         - Verify devices patches inventory
+         - Perform a noop rootfs update and verify the update was successful
+        """
+        gateway_addr = env.get_mender_gateway()
+        URL_MGMT_USERADM = useradm.URL_MGMT.replace("mender-api-gateway", gateway_addr)
+        URL_MGMT_DEVAUTH = deviceauth.URL_MGMT.replace(
+            "mender-api-gateway", gateway_addr
+        )
+        URL_MGMT_INVNTRY = inventory.URL_MGMT.replace(
+            "mender-api-gateway", gateway_addr
+        )
+        URL_MGMT_DPLMNTS = deployments.URL_MGMT.replace(
+            "mender-api-gateway", gateway_addr
+        )
```

Comment:
Wouldn't it be nice that `testutils.client.GATEWAY_URL` could be parametrized? But then, maybe that is that it lead to the current `MenderAPI` class with global scope...

Reply:
This is something I would like to discuss further. Here in the test is okey, but putting a similar `replace` in `testutils.common.create_org` is rather hacky.

It goes all the way to the base design of `client.ApiClient` class, where the full url is expected to be passed to the constructor, but then these URLs are taken from the micro service itself  prepending a GATEWAY_URL that comes from the `client`. So this is a circular dependency that we should break.

A backwards compatible solution could be to add some methods in the micro services API files like:

```
def url_mgmt(base_url=testutils.api.client.GATEWAY_URL):
    return base_url + "/api/management/v1/useradm"

URL_MGMT = management_url()
```

So that backend-tests could use `URL_MGMT` while integration tests can use `url_mgmt(custom_gateway)`.

This is also an step forward for backend-test in staging environment project ;)

Reply:
Yeah, this is pretty ugly. Maybe we could reduce the `URL_{MGMT,DEVICE}` to only hold the base url path and let the `gateway_host` be a separate argument (defaulting to "mender-api-gateway") when initializing the ApiClient class?

Reply:
That is also a valid idea and will result in a cleaner interface, I think.

You would then need to modify integration test `test_os_ent_migration` [here](https://github.com/mendersoftware/integration/blob/master/tests/tests/test_os_ent_migration.py#L92).